redwood-protocol/src/commonMain/kotlin/app/cash/redwood/protocol/protocol.kt

```diff
@@ -50,13 +57,37 @@ public data class PropertyDiff(
 public data class LayoutModifiers(
   /** Identifier for the widget whose layout modifier has changed. */
   val id: Id,
-  /**
-   * Array of layout modifiers. Each element of this array is itself a two-element array of the
-   * layout modifier tag and then serialized value.
-   */
-  val elements: JsonArray,
+  val elements: List<LayoutModifierElement> = emptyList(),
 )
 
+@Serializable(with = LayoutModifierElementSerializer::class)
+public data class LayoutModifierElement(
+  val tag: LayoutModifierTag,
+  val value: JsonElement = JsonNull,
+)
+
+private object LayoutModifierElementSerializer : KSerializer<LayoutModifierElement> {
+  private val delegate = serializer<Array<JsonElement>>()
+
+  @OptIn(ExperimentalSerializationApi::class)
+  override val descriptor = SerialDescriptor("LayoutModifierElement", delegate.descriptor)
+
+  override fun deserialize(decoder: Decoder): LayoutModifierElement {
+    val decoded = decoder.decodeSerializableValue(delegate)
+    check(decoded.size in 1..2) { "TODO" }
```

Comment:
lenient?

```
check(decoded.size >= 2)
```

Reply:
0 elements are also invalid. The tag is required.

But I'll add the message I forgot and maybe write a test for this logic, too.